src/jobserver_test.cc

```diff
@@ -0,0 +1,345 @@
+  /// created external token pool.
+  void ServerCreate(const char* name, size_t count);
+
+  /// Return the number of tokens currently available in the external token
+  /// pool.
+  size_t ServerCount();
+
+  /// Remove/close the handle to external token pool.
+  void ServerRemove();
+
+  /// Wrapped jobserver object to test on.
+  JobserverTest jobserver_;
+
+  /// Stored makeflags read before starting tests.
+  const char* makeflags_ = nullptr;
+
+  /// Name of created external jobserver token pool.
+  const char* name_ = nullptr;
+
+#ifdef _WIN32
+  /// Implementation of posix setenv() for windows that forwards calls to
+  /// _putenv().
+  int setenv(const char* name, const char* value, int _) {
+    std::string envstring;
+
+    // _putenv() requires a single <name>=<value> string as argument.
+    envstring += name;
+    envstring += '=';
+    envstring += value;
+
+    return _putenv(envstring.c_str());
+  };
+
+  /// Implementation of posix unsetenv() for windows that forwards calls to
+  /// _putenv().
+  int unsetenv(const char* name) {
+    /// Call _putenv() with <name>="" to unset the env variable.
+    return setenv(name, "", 0);
+  }
+
+  /// Handle of the semaphore used as external token pool.
+  HANDLE sem_ = INVALID_HANDLE_VALUE;
+#else
+  /// File descriptor of the fifo used as external token pool.
+  int fd_ = -1;
+#endif
+};
+
+JobserverTestState::JobserverTestState() {
+  makeflags_ = getenv("MAKEFLAGS");
+  unsetenv("MAKEFLAGS");
+}
+
+JobserverTestState::~JobserverTestState() {
+  if (name_ != nullptr) {
+    ServerRemove();
+  }
+
+  if (makeflags_ != nullptr) {
+    setenv("MAKEFLAGS", makeflags_, 1);
+  } else {
+    unsetenv("MAKEFLAGS");
+  }
+}
+
+void JobserverTestState::ServerConfigure(const char* name)
+{
+  std::string makeflags("--jobserver-auth=");
+#ifdef _WIN32
+  makeflags += "sem:";
+#else
+  makeflags += "fifo:";
+#endif
+  makeflags += name;
+
+  assert(setenv("MAKEFLAGS", makeflags.c_str(), 1) == 0);
+
+}
+
+#ifdef _WIN32
+void JobserverTestState::ServerCreate(const char* name, size_t count) {
+  assert(name_ == nullptr);
+  ServerConfigure(name);
+  name_ = name;
+
+  // One cannot create a semaphore with a max value of 0 on windows
+  sem_ = CreateSemaphoreA(nullptr, count, count ? count : 1, name);
+  assert(sem_ != nullptr);
+}
+
+size_t JobserverTestState::ServerCount() {
+  assert(name_ != nullptr);
+  size_t count = 0;
+
+  // First acquire all the available tokens to count them
+  while (WaitForSingleObject(sem_, 0) == WAIT_OBJECT_0) {
+    count++;
+  }
+
+  // Then return the acquired tokens to let the client continue
+  ReleaseSemaphore(sem_, count, nullptr);
+
+  return count;
+}
+
+void JobserverTestState::ServerRemove() {
+  assert(name_ != nullptr);
+  CloseHandle(sem_);
+  name_ = nullptr;
+}
+
+#else // _WIN32
+
+void JobserverTestState::ServerCreate(const char* name, size_t count) {
+  assert(name_ == nullptr);
```

Comment:
Shouldn't this be `name_ != nullptr` ?

In general it is best to avoid assert() statements in tests. The alternatives are:

- Use ASSERT_XXX() macros directly inside a function that returns no value, such as this one. This translates into a `return` statement but allows you to print decent information in case of error with `<< some_message` syntax.

- Write functions that return bool on success, and ASSERT_TRUE(func(...)) at the call site. This is simple but does not give many details when an error occurs.

-

Reply:
The `assert()` ensures that `ServerCreate()` isn't called multiple times without the corresponding call to `ServerRemove()` in between.

I'll replace the `assert()` calls with `ASSERT_XX()` or `return true|false` as appropriate.

Reply:
Please do :-) The `ASSERT_XXX()` also provides a nice clean exit that ensures that the test destructor is called, cleaning up any system state if needed, while `assert()` just crashes (when enabled), leaving all kinds of temporary files on your system sometimes.

Reply:
Yeah, tests were failing because `assert()` statements are no-ops in release builds... Another good reason not to use them :)